### DISCUSSION TITLE
Exists

### DISCUSSION THREAD

**Discussion by vulcandragi** [SeaQL/sea-orm#730] [Ideas]
Add a variant of `find`, which checks if the record exists in the database but doesn't return any data, just says whether or not it exists

This improves performance in some cases where a test of whether the user has already been registered and the case will not be registered

There are better ways to do this test, which is in the insertion case of failure, check the error code, but as it is not so easy at the moment to obtain this code, it would be easier to implement this function

**Comment by RomainMazB** [comment]
This would be useful indeed, I may take a look to implement this in the next few days.

**Comment by billy1624** [comment]
How about adding a new operation `exists` under find?

```rust
let _: bool = Entity::find_by_id(1)
    .exists(db)
    .await?

let _: bool = Entity::find()
    .filter( ... )
    .exists(db)
    .await?
```

**Reply by cemremengu** [reply]
Was this ever implemented or still up for a PR?

**Comment by vulcandragi** [comment]
I think it's good like this, just one thing that maybe you should take into account, when using the filter maybe there can be more than one, but when the filter is by id, it can only exist or not

So maybe you can use `count(1)` in the filter to be able to nab the amount of existing records

**Comment by billy1624** [comment]
Btw... you could also `count` it with paginator
https://docs.rs/sea-orm/0.4.2/sea_orm/trait.PaginatorTrait.html#method.count

**Comment by vulcandragi** [comment]
yes, I agree with what you said, but that would be one more call to the database

that's why I talked about the filter returning the number of records, but it's just my opinion

**Comment by RomainMazB** [comment]
🤔 Using the sql `EXISTS` statement, no matter if there is one or multiple results, it will just return `TRUE` or `FALSE` directly from the database.

I thought this would be already implemented into sea-query but it's not. I guess this should be the starting point.

Related to https://github.com/SeaQL/sea-query/issues/118

**Comment by vulcandragi** [comment]
makes sense, and when I did that I didn't use this command, but looking at it from this side it only makes sense to return the boolean value

**Comment by tyt2y3** [comment]
I am still thinking whether we should add the method to `PaginatorTrait` or create a new `Trait` (which includes other methods like `max` or `min`)

**Comment by vulcandragi** [comment]
I think it would be interesting to be separated as it doesn't have anything to do strictly with pagination, but something useful that can be used in any context.

**Comment by penso** [comment]
I'm currently slowly switch a Ruby/ActiveRecord API to Rust and would love to see an `exists(db)` returning true/false. The Ruby AR query looks like this: `SELECT 1 AS one FROM your_model LIMIT 1` which could be the most effective.

**Reply by ikrivosheev** [reply]
Related to: https://github.com/SeaQL/sea-query/issues/118

**Comment by Xiaobaishushu25** [comment]
Can the `exists` function be used in seaorm now? Do you have any simple examples?

**Reply by tyt2y3** [reply]
You can use the `async fn count()` from `PaginatorTrait`.

**Comment by IgnisDa** [comment]
@billy1624 @tyt2y3 Now that it has been merged to sea-query, can we expect an `.exists()` function? `.count()` is not as performant as exists.

**Reply by Expurple** [reply]
I can't say on their behalf, but I'm a new co-mainainer and we're currently working on the next major release of SeaORM and SeaQuery. So, I prioritize breaking-change ideas and PRs higher. I assume that this will go on for a couple more months.

I have definitely wished for `.exists()` many times in my own codebase, and this feature is on my own long-term todo list.

TL;DR we'd like to have this, but without outside contributions it probably won't happen earlier than months from now.

**Comment by IgnisDa** [comment]
If you could tell me where you'd like it implemented, I'd love to make a PR
for it.

**Reply by Expurple** [reply]
I'm not familiar with this area myself. But I'd start with looking at the similar `async` SeaORM methods that consume the statement and execute it in the provided connection/transaction. Methods like `count`, `one`, `all`.

[`count`](https://docs.rs/sea-orm/1.1.12/sea_orm/trait.PaginatorTrait.html#method.count) is in [`PaginatorTrait`](https://docs.rs/sea-orm/1.1.12/sea_orm/trait.PaginatorTrait.html).

For some reason, `one` and `all` are duplicated inherent methods in [`Selector`](https://docs.rs/sea-orm/1.1.12/sea_orm/struct.Selector.html#method.one), [`SelectorRaw`](https://docs.rs/sea-orm/1.1.12/sea_orm/struct.SelectorRaw.html#method.one), and [`Select*`](https://docs.rs/sea-orm/1.1.12/sea_orm/query/struct.SelectTwo.html#method.one) types. This looks like a design flaw and something that should be abstracted into a trait. Added this to https://github.com/SeaQL/sea-orm/discussions/2548

**Reply by IgnisDa** [reply]
@Expurple Made an initial PR in #2623. Would appreciate your feedback on it.

**Reply by IgnisDa** [reply]
Hey @Expurple , is there any ETA on when this could be merged? Thanks!

**Comment by Expurple** [comment]
Merged in #2623
